File: csharp/demo/demo12c01-race-cond-data-race.cs
/*
 * RACE CONDITIONS AND DATA RACES
 */
using System;
using System.Collections.Generic;
using System.Threading;



class Demo12C01 : IRunnable
{
    private int counter;


    public void run()
    {
        const int NUM_THREADS = 16;

        counter = 0;

        var lstTh = new List<Thread>();

        for (int i = 0; i < NUM_THREADS; ++i)
        {
            lstTh.Add(new Thread(doTask));
        }

        lstTh.ForEach(th => th.Start());
        lstTh.ForEach(th => th.Join());

        Console.WriteLine("counter = " + counter);
        // We are NOT sure that counter = 16000
    }


    private void doTask()
    {
        Thread.Sleep(1000);

        for (int i = 0; i < 1000; ++i)
            ++counter;
    }
}

File: python/demo12c01_race_cond_data_race.py
'''
RACE CONDITIONS AND DATA RACES
'''

import time
import threading



counter = 0



def do_task():
    global counter

    for _ in range(1000):
        temp = counter + 1
        time.sleep(0.0001)
        counter = temp



lstth = [threading.Thread(target=do_task) for _ in range(32)]

for th in lstth:
    th.start()

for th in lstth:
    th.join()

print('counter =', counter)
# We are NOT sure that counter = 32000
